docs/design/reserve-action.md

```diff
@@ -0,0 +1,152 @@
+    + [Job starving](#job-starving)
+    + [Node resource reservation rule](#node-resource-reservation-rule)
+  * [Data Model](#data-model)
+  
+## Introduction
+
+The `reserve` action is to solve the problem that the task can not be scheduled for a long time. We call these long 
+unscheduled jobs starving job, if a job is not scheduled for a long time, it is a starving job. The `reserve` action 
+will try to reserve some node resources for the pods under the starving job. Nodes who has reserved resources will no 
+longer schedule new pods. When the existing pods on the node are completed, the node resources they occupy will be 
+released. Pods under the starving job will scheduled on such nodes.
+
+## Motivation
+
+When the node resources in the cluster are insufficient, the tasks with large resource requirements will be at risk of 
+starvation. The so-called starvation means that the task can not be scheduled for a long time due to various reasons.
+
+There are many reasons why a pod can not be scheduled. From the perspective of task resource requirement and cluster 
+allocatable resources, if the resources required by the task is larger that the allocatable resource of any node in the 
+cluster, the task can not be scheduled. The node under the cluster needs to be expanded vertically to increase the node 
+resources.
+
+If the node resources in the cluster has meet the resource requirements of the pod, but because of that the resources of 
+the nodes are occupied by pods with relatively small resource requirements for a long time, the task scheduling of large
+resource requirement can not be satisfied. In this case, there is no need to expand the node resources, and the `reserve` 
+action can solve such problems.
+
+The `reserve` action prevents the large resource requirement task from being starved by reserving the node resources for
+it and not allowing the other tasks to be scheduled on the node.
+
+## Function Detail 
+
+### Job starving
+
+If the `pending` time of job is greater that the set time threshold, the task has the risk of starvation, and needs to
+reserve node resources for it. If there is unscheduled task under the job, but the job has already met the `running` 
+condition and is already `running`, we think that this job does not need to reserve node resource any more.
+
+### Node resource reservation rule
+
+If the allocatable resources on the node is greater that the resource request of the task, then the node satisfies the 
+scheduled requirement of the task now or later. This node can reserve resources for this task. If the idle resource of 
+the current node has met the scheduling condition of the task, resources will be allocated to this task, and the task 
+will be directly assigned to the node. If the idle resource of the node can not meet the resource requirement of the task,
+the node will reserve resources for this task. Once a node reserve resources for a task, no other tasks will be scheduled 
```

Comment:
If two tasks require affinity, then what will happen?

Reply:
L57 should answer me

Reply:
The `PredicateFn` will be called during selecting the reserved node. For a pair of tasks configured with affinity, after the first task reserves resources on the node successfully, the result of `PredicateFn` for the second task will be the node who reserve resources for the first task. If the node can also reserve resources for the second task, the second task will be reserved on this node, otherwise the sencond task will be failed to reserve resources in this round. Regardless of the results of the above two tasks, for the job of these two tasks, if the state of the job can be ready, the reservation of the node for tasks will be performed, otherwise, the reservation will be discarded.